rollouts/controllers/remoterootsync_controller.go

```diff
@@ -145,42 +156,65 @@ func (r *RemoteRootSyncReconciler) Reconcile(ctx context.Context, req ctrl.Reque
-	dynCl, err := r.getDynamicClientForCluster(ctx, clusterRef)
-	if err != nil {
-		return ctrl.Result{}, err
-	}
+	syncStatus, syncError := r.syncExternalSync(ctx, &remoterootsync)
 
-	if err := r.patchRootSync(ctx, dynCl, req.Name, &remoterootsync); err != nil {
+	if err := r.updateStatus(ctx, &remoterootsync, syncStatus, syncError); err != nil {
+		logger.Error(err, "Failed to update status")
 		return ctrl.Result{}, err
 	}
 
-	r.setupWatches(ctx, remoterootsync.Name, remoterootsync.Namespace, remoterootsync.Spec.ClusterRef)
+	return ctrl.Result{}, syncError
+}
+
+func (r *RemoteRootSyncReconciler) syncExternalSync(ctx context.Context, rrs *gitopsv1alpha1.RemoteRootSync) (string, error) {
+	syncName := rrs.Name
+	clusterRef := &rrs.Spec.ClusterRef
 
-	syncStatus, err := checkSyncStatus(ctx, dynCl, req.Name)
+	dynCl, err := r.getDynamicClientForCluster(ctx, clusterRef)
 	if err != nil {
-		return ctrl.Result{}, err
+		return "", fmt.Errorf("failed to create client: %w", err)
 	}
 
-	if err := r.updateStatus(ctx, &remoterootsync, syncStatus); err != nil {
-		logger.Error(err, "Failed to update status")
-		return ctrl.Result{}, err
+	if err := r.patchRootSync(ctx, dynCl, syncName, rrs); err != nil {
+		return "", fmt.Errorf("failed to create/update sync: %w", err)
+	}
+
+	r.setupWatches(ctx, rrs.Name, rrs.Namespace, rrs.Spec.ClusterRef)
+
+	syncStatus, err := checkSyncStatus(ctx, dynCl, syncName)
+	if err != nil {
+		return "", fmt.Errorf("faild to check status: %w", err)
 	}
 
-	return ctrl.Result{}, nil
+	return syncStatus, nil
 }
 
-func (r *RemoteRootSyncReconciler) updateStatus(ctx context.Context, rrs *gitopsv1alpha1.RemoteRootSync, syncStatus string) error {
+func (r *RemoteRootSyncReconciler) updateStatus(ctx context.Context, rrs *gitopsv1alpha1.RemoteRootSync, syncStatus string, syncError error) error {
 	logger := klog.FromContext(ctx)
 
-	// Don't update if there are no changes.
-
 	rrsPrior := rrs.DeepCopy()
+	conditions := &rrs.Status.Conditions
 
-	rrs.Status.SyncStatus = syncStatus
-	rrs.Status.ObservedGeneration = rrs.Generation
+	if syncError == nil {
+		rrs.Status.SyncStatus = syncStatus
 
-	meta.SetStatusCondition(&rrs.Status.Conditions, metav1.Condition{Type: externalSyncCreatedConditionType, Status: metav1.ConditionTrue, Reason: "SyncCreated"})
+		meta.RemoveStatusCondition(conditions, conditionReconciling)
+		meta.RemoveStatusCondition(conditions, conditionStalled)
+	} else {
+		reconcileReason := reasonUpdateSync
+
+		rrs.Status.SyncStatus = "Unknown"
+
+		if !r.isExternalSyncCreated(rrs) {
+			rrs.Status.SyncStatus = ""
+			reconcileReason = reasonCreateSync
+		}
+
+		meta.SetStatusCondition(conditions, metav1.Condition{Type: conditionReconciling, Status: metav1.ConditionTrue, Reason: reconcileReason})
+		meta.SetStatusCondition(conditions, metav1.Condition{Type: conditionStalled, Status: metav1.ConditionTrue, Reason: reasonError, Message: syncError.Error()})
+	}
+
+	rrs.Status.ObservedGeneration = rrs.Generation
```

Comment:
I wonder if we can simplify (combining `syncStatus` with other conditions). So, a RRS can be in four states: `{Created, Reconciling, Stalled, Ready}` and state can transition between these four or am I thinking in the wrong way ?

Reply:
My knowledge is still evolving here, but the `Reconciling` and `Stalled` conditions are from [kstatus conditions](https://github.com/kubernetes-sigs/cli-utils/tree/master/pkg/kstatus#conditions) so these do have specific definitions, namely if the Rollout controller has reconciled its changes to the resources it directly owns. In this case, it is if the RootSync the RemoteRootSync owns exists and is up to date with the latest spec (regardless of the status of the RootSync resource itself), so this would deviate from that pattern which the kstatus library depends on.

Beyond that, there's no other reason we could have a custom condition type where the four states could be captured with Reason.

@mortent thoughts?

Reply:
Thanks for the detailed answer here. Looked at the logic again:

I think `Reconciled`, `Stalled` is enough to convey the status to the end-users. `Ready` is not needed for that. So this is good so far.
However the current logic to determine whether the controller created remote RSync object or not  seems a bit unintuitive implemented using the reason field in `reconciling` condition. If we think more it is is essentially a state that is for internal purpose to be used while finalizing and it is not really useful for the end-user status. Seems like having separate field in the status to track `RSyncObjectCreatedOrNot` would have kept the implementation cleaner. WDYT ?

Reply:
Added `Status.SyncCreated` boolean field with the latest commit. This eliminates the need for the isExternalSyncCreated function.